src/includes_normalize-win32.cc

```diff
@@ -38,34 +76,54 @@ bool SameDrive(StringPiece a, StringPiece b)  {
   return _stricmp(a_drive, b_drive) == 0;
 }
 
-}  // anonymous namespace
+bool IsAbsPath(StringPiece s) {
+  if (s.size() < 3 ||
+      !isalpha(s[0]) ||
+      s[1] != ':' ||
+      !IsPathSeparator(s[2])) {
+    return false;
+  }
+
+  // Check "." or ".." is contained in path.
+  for (size_t i = 2; i < s.size(); ++i) {
+    if (!IsPathSeparator(s[i])) {
+      continue;
+    }
 
-string IncludesNormalize::Join(const vector<string>& list, char sep) {
-  string ret;
-  for (size_t i = 0; i < list.size(); ++i) {
-    ret += list[i];
-    if (i != list.size() - 1)
-      ret += sep;
+    // Check ".".
+    if (i + 1 < s.size() && s[i+1] == '.' &&
+        (i + 2 >= s.size() || IsPathSeparator(s[i+2]))) {
+      return false;
+    }
+
+    // Check "..".
+    if (i + 2 < s.size() && s[i+1] == '.' && s[i+2] == '.' &&
+        (i + 3 >= s.size() || IsPathSeparator(s[i+3]))) {
+      return false;
+    }
   }
-  return ret;
-}
 
-vector<string> IncludesNormalize::Split(const string& input, char sep) {
-  vector<string> elems;
-  stringstream ss(input);
-  string item;
-  while (getline(ss, item, sep))
-    elems.push_back(item);
-  return elems;
+  return true;
 }
 
-string IncludesNormalize::ToLower(const string& s) {
-  string ret;
-  transform(s.begin(), s.end(), back_inserter(ret), ::tolower);
-  return ret;
+}  // anonymous namespace
+
+IncludesNormalize::IncludesNormalize(const string& relative_to) {
+  relative_to_ = AbsPath(relative_to);
+  splitted_relative_to_ = SplitStringPiece(relative_to_, '/');
 }
 
 string IncludesNormalize::AbsPath(StringPiece s) {
+  if (IsAbsPath(s)) {
```

Comment:
Most of slowness is come from GetFullPathName API.
Also `s` is often FullPath style here.

Stats without this pruning
```
metric                  count   avg (us)        total (ms)
.ninja parse            2771    1238.2          3431.0
canonicalize str        701014  0.0             0.6
canonicalize path       10164351        0.0             54.6
lookup node             9391196 0.0             197.2
.ninja_log load         1       81460.0         81.5
.ninja_deps load        1       301154.0        301.2
node stat               83322   6.5             540.4
depfile load            896     24.0            21.5
StartEdge               28284   4042.6          114341.6
FinishCommand           28283   3866.6          109359.7
CLParser::Parse         21228   3060.4          64966.0

path->node hash load 0.93 (122492 entries / 131072 buckets)

real    4m48.999s
user    0m0.000s
sys     0m0.000s
```

Stats with this pruning
```
metric                  count   avg (us)        total (ms)
.ninja parse            2771    1149.5          3185.4
canonicalize str        701014  0.0             5.2
canonicalize path       10164351        0.0             48.6
lookup node             9424852 0.0             94.8
.ninja_log load         1       75764.0         75.8
.ninja_log recompact    1       356286.0        356.3
.ninja_deps load        1       211452.0        211.5
node stat               83322   6.1             511.4
depfile load            896     20.4            18.3
StartEdge               28284   3988.5          112811.3
FinishCommand           28283   2961.1          83749.0
CLParser::Parse         21228   2175.9          46190.2
 
path->node hash load 0.93 (122492 entries / 131072 buckets)
 
real    4m21.726s
user    0m0.000s
sys     0m0.016s
```

I feel this pruning makes build lot faster.